File: src/erlang/commstime.erl
-module(commstime).
-export([start/0, commstime/2, printer/1, delta/1, succ/1, prefix/2]).

id(Out, Count) ->
  receive N -> Out ! N end,
  if
    Count > 0 -> id(Out, Count - 1);
    true -> ok
  end.

prefix(N, Count) ->
  receive Out -> Out ! N end,
  id(Out, Count).

save_results([], File) ->
  ok;
save_results([X|Results], File) ->
  file:write(File, io_lib:fwrite("~p\n", [X])),
  save_results(Results, File).

printer(Count, Total, Results) ->
  if
    Count rem 10000 == 0 ->
      {_, Time} = statistics(wall_clock),
      receive _ ->
        if
          Total > 1 -> printer(1, Total - 1, [Time / 10000|Results]);
          true ->
            {ok, File} = file:open("ct-erl.csv", [write]),
            save_results(Results, File)
        end
      end;
    true ->
      receive _ -> printer(Count + 1, Total - 1, Results) end
  end.

printer(Total) ->
  statistics(wall_clock),
  printer(1, Total, []).

delta(Out0, Out1, Total) ->
  receive N -> Out0 ! N, Out1 ! N end,
  if
    Total > 0 -> delta(Out0, Out1, Total - 1);
    true -> ok
  end.

delta(Total) ->
  receive {Out0, Out1} -> delta(Out0, Out1, Total) end.

succ(Out, Total) ->
  receive N -> Out ! (N + 1) end,
  if
    Total > 0 -> succ(Out, Total - 1);
    true -> ok
  end.

succ(Total) ->
  receive Out -> succ(Out, Total) end.

commstime(N, Total) ->
  Delta = spawn(commstime, delta, [Total]),
  Succ = spawn(commstime, succ, [Total]),
  Pre = spawn(commstime, prefix, [N, Total]),
  Delta ! {self(), Succ},
  Pre ! Delta,
  Succ ! Pre,
  printer(Total).

start() ->
  commstime(0, 10000000).

File: src/go/commstime.go
// Author: Kevin Chalmers
// Date: 17/08/2017

package main

import "fmt"
import "time"
import "os"

// id sends on out whatever is received on in.
func id(in chan int, out chan int, count int) {
	for index := 0; index < count; index++ {
		x := <-in
		out <- x
	}
}

// prefix outputs N, then behaves as id.
func prefix(N int, in chan int, out chan int, count int) {
	out <- N
	id(in, out, count)
}

// printer reads in 10000 values and calculates the time taken.  Results are
// stored in ct-go.csv.
func printer(in chan int, count int) {
	var results [100]int64
	start := time.Now()
	for index := 0; index < count; index++ {
		if index%10000 == 0 {
			total := time.Now().Sub(start)
			results[index/10000] = total.Nanoseconds() / 10000
			fmt.Println(results[index/10000])
			start = time.Now()
		}
		<-in
	}
	f, err := os.Create("ct-go.csv")
	if err != nil {
		fmt.Println("error creating file: ", err)
		return
	}
	defer f.Close()
	for index := 0; index < 100; index++ {
		_, err = f.WriteString(fmt.Sprintf("%d\n", results[index]))
		if err != nil {
			fmt.Println("error writing string: ", err)
		}
	}
}

// succ reads a value from in, increments it, and outputs on out.
func succ(in chan int, out chan int, count int) {
	for index := 0; index < count; index++ {
		x := <-in
		out <- x + 1
	}
}

// delta reads from in, then outputs on out0 then out1.
func delta(in chan int, out0 chan int, out1 chan int, count int) {
	for index := 0; index < count; index++ {
		x := <-in
		out0 <- x
		out1 <- x
	}
}

func main() {
	count := 10000 * 100
	a := make(chan int)
	b := make(chan int)
	c := make(chan int)
	d := make(chan int)
	go prefix(0, a, b, count)
	go delta(b, c, d, count)
	go succ(c, a, count)
	printer(d, count)
}
